integration/integration_test.go

```diff
@@ -0,0 +1,463 @@
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"testing"
+)
+
+var integration = flag.Bool("integration", false, "run integration tests")
+
+// image such as docker.io/kinvolk/gadget:latest
+var image = flag.String("image", "", "gadget container image")
+
+var githubCI = flag.Bool("github-ci", false, "skip some tests which cannot be run on GitHub CI due to host kernel not BPF ready")
+
+func runCommands(cmds []*command, t *testing.T) {
+	failed := false
+
+	for _, cmd := range cmds {
+		failed = cmd.run(t, failed)
+	}
+}
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+
+	if !*integration {
+		fmt.Println("Skipping integration test.")
+
+		os.Exit(0)
+	}
+
+	if os.Getenv("KUBECTL_GADGET") == "" {
+		fmt.Fprintf(os.Stderr, "please set $KUBECTL_GADGET.")
+
+		os.Exit(-1)
+	}
+
+	if *image != "" {
+		os.Setenv("GADGET_IMAGE_FLAG", "--image "+*image)
+	}
+
+	initCommands := []*command{
+		deployInspektorGadget,
+		waitUntilInspektorGadgetPodsDeployed,
+		waitUntilInspektorGadgetPodsInitialized,
+		createTestNamespace,
+	}
+
+	fmt.Printf("Setup inspektor-gadget:\n")
+	for _, cmd := range initCommands {
+		err := cmd.runWithoutTest()
+
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "%v\n", err)
+
+			os.Exit(-1)
+		}
+	}
+
+	ret := m.Run()
+
+	fmt.Printf("Clean inspektor-gadget:\n")
+	cleanupTestNamespace.runWithoutTest()
+	cleanupInspektorGadget.runWithoutTest()
+
+	os.Exit(ret)
+}
+
+func TestBindsnoop(t *testing.T) {
+	bindsnoopCmd := &command{
+		name:           "Start bindsnoop gadget",
+		cmd:            "$KUBECTL_GADGET bindsnoop -n test-ns",
+		expectedRegexp: `test-ns\s+test-pod\s+test-pod\s+\d+\s+nc`,
+		startAndStop:   true,
+	}
+
+	commands := []*command{
+		bindsnoopCmd, // Start it.
+		{
+			name:           "Run pod which calls bind()",
+			cmd:            "kubectl run --restart=Never --image=busybox -n test-ns test-pod -- sh -c 'while true; do nc -l -p 9090 -w 1; done'",
```

Comment:
nit: I think the `while true` here is not needed as `nc -l -p 9090` will block.

Reply:
`-w 1` options of `nc` will make it exits after one second.
Thus, I prefer to keep it and to run several of them to be sure gadget output is populated.

Reply:
Actually -w doesn't make sense for -l:

from the man page:

```
-w timeout
             Connections which cannot be established or are idle timeout after timeout seconds.  The -w flag has no effect on the -l option, i.e. nc will
             listen forever for a connection, with or without the -w flag.  The default is no timeout.
```

You can keep the while, it's not a big deal anyway.

Reply:
I cannot confirm as I use `busybox`'s `nc` and the documentation sadly does not detail this:
> -w SEC Timeout for connects and final net reads